### DISCUSSION TITLE
Can we persist parsed tree into disk and load it back later?

### DISCUSSION THREAD

**Discussion by QthCN** [tree-sitter/tree-sitter#1237] [Q&A] [answered]
Hi,

We are working on a repository analysis project and the project needs to analysis all git blobs AST generated by tree-sitter, for some large git repository it may contains more than one million blobs. We found it takes a lot of time on the parsing so we are thinking about to persist parsed result into LMDB(a mmap supported kv storage) and we can create a cursor on the mapped memory to operate the parsed AST tree directly.

I search the document and didn't found any API support the `tree dump` or `create cursor from memory block`. So I create this discussion to get some help on how to implement it.

Thanks.

**Comment by ahlinc** [comment, marked as answer]
Tree-sitter doesn't have such functionality, it's internal in-memory tree representation is pretty complex. Starting point to get into:
https://github.com/tree-sitter/tree-sitter/blob/ad30779d90c51f8f27bc648254cb16d83ab1ab1d/lib/src/subtree.h#L108-L112

If you don't need to use tree-sitter queries on the stored AST and can solve your analysis task by traversing the AST than it would be much easier to implement own serialisation on top of tree-sitter's tree by traversing it after parsing and after that do all own analysis by traversing on the own AST representation.

Also note:
* Tree-sitter doesn't store real input text in tree nodes and stores only text range references for every node.
* It may store an external scanner state for many nodes that overall may take significant amount of memory and the storage in your case if to try to dump the entire tree-sitter representation. What may be a wasted storage space if you don't need to reparse blobs with some edits in them.

**Comment by razzeee** [comment]
Also see https://github.com/tree-sitter/tree-sitter/issues/824 for reference
